cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+                    self._mm.apply_swap(*swap)
+                process_executable_ops(idx)
+
+        # edge case: single qubit gates may appear in moments after all 2-qubit gates
+        for i in range(len(timesteps), len(single_qubit_ops)):
+            routed_ops.append([self._mm.mapped_op(op) for op in single_qubit_ops[i]])
+
+        return routed_ops
+
+    def _disjoint_qubit_combinations(
+        self, candidates: List[Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]]
+    ) -> List[Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]]:
+        single_candidates = [x[0] for x in candidates]
+        return [
+            pair
+            for pair in combinations(single_candidates, 2)
+            if set(q for q in pair[0]).isdisjoint(set(q for q in pair[1]))
+        ]
+
+    def _symmetry_swap_pair(
+        self, timesteps: List[List['cirq.Operation']], idx: int, max_search_radius: int
+    ) -> Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]:
+        """Computes cost function with pairs of candidate swaps that act on disjoint qubits."""
+        pair_sigma = self._disjoint_qubit_combinations(
+            self._initial_candidate_swaps(timesteps[idx])
+        )
+        for s in range(idx, min(max_search_radius + idx, len(timesteps))):
+            if len(pair_sigma) <= 1:
+                break
+            pair_sigma = self._next_candidate_swaps(pair_sigma, timesteps[s])
+
+        if len(pair_sigma) > 1 and idx + max_search_radius <= len(timesteps):
+            return self._symmetry_brute_force(timesteps, idx)
+        chosen_swap_pair = pair_sigma[0]
+        return (chosen_swap_pair[0], chosen_swap_pair[1])
+
+    def _symmetry_brute_force(
+        self, timesteps: List[List['cirq.Operation']], idx: int
+    ) -> Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]:
+        """Inserts SWAPS along the shortest path of the qubits that are the farthest."""
+        qubits = max(
+            [(op.qubits, self._mm.dist_on_device(*op.qubits)) for op in timesteps[idx]],
+            key=lambda x: x[1],
+        )[0]
+        path = self._mm.shortest_path(*qubits)
+        q1 = path[0]
+        return tuple([(q1, path[i + 1]) for i in range(len(path) - 2)])
+
+    def _initial_candidate_swaps(
+        self, timestep_ops: List['cirq.Operation']
+    ) -> List[Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]]:
+        """Finds all feasible SWAPs between qubits involved in 2-qubit operations."""
+        physical_qubits = set(self._mm.map[op.qubits[i]] for op in timestep_ops for i in range(2))
+        physical_swaps = self._mm.induced_subgraph.edges(nbunch=physical_qubits)
+        return [
+            ((self._mm.inverse_map[q1], self._mm.inverse_map[q2]),) for q1, q2 in physical_swaps
+        ]
+
+    def _next_candidate_swaps(
+        self,
+        candidate_swaps: List[Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]],
+        timestep_ops: List['cirq.Operation'],
+    ) -> List[Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]]:
```

Comment:
```suggestion
    ) -> List[QidPair]:
```

Reply:
In order to support ranking swaps and pairs of swaps it needs to be `List[Tuple[QidPair, ...]]`